File: app/src/main/java/com/eftimoff/bakingapp/recipelist/view/RecipeListFragment.java
package com.eftimoff.bakingapp.recipelist.view;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.Toast;

import com.eftimoff.bakingapp.R;
import com.eftimoff.bakingapp.app.injection.AppComponent;
import com.eftimoff.bakingapp.app.models.Recipe;
import com.eftimoff.bakingapp.app.view.BaseFragment;
import com.eftimoff.bakingapp.recipedetails.view.RecipeDetailsActivity;
import com.eftimoff.bakingapp.recipelist.di.RecipeListModule;
import com.eftimoff.bakingapp.recipelist.presenter.RecipeListPresenter;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;

public class RecipeListFragment extends BaseFragment implements RecipeListView, RecipeAdapter.Callback {

    @BindView(R.id.recipeList)
    RecyclerView recipeList;

    @Inject
    RecipeListPresenter presenter;
    @Inject
    RecipeAdapter recipeAdapter;
    @Inject
    RecyclerView.LayoutManager layoutManager;

    public static RecipeListFragment newInstance() {
        return new RecipeListFragment();
    }

    @Override
    protected int getLayoutRes() {
        return R.layout.fragment_recipe_list;
    }

    @Override
    protected void inject(AppComponent appComponent) {
        appComponent.plus(new RecipeListModule(this)).inject(this);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        recipeAdapter.setCallback(this);
        recipeList.setAdapter(recipeAdapter);
        recipeList.setLayoutManager(layoutManager);

        presenter.loadRecipes();
    }

    @Override
    public void showRecipeList(List<Recipe> recipes) {
        recipeAdapter.setRecipes(recipes);
    }

    @Override
    public void showRecipeError() {
        Toast.makeText(getContext(), "Recipe error while fetching from backend.", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onRecipeClicked(Recipe recipe) {
        RecipeDetailsActivity.start(getContext(), recipe);
    }
}

File: app/src/main/java/com/eftimoff/bakingapp/recipelist/view/RecipeListFragment.kt
package com.eftimoff.bakingapp.recipelist.view

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v7.widget.RecyclerView
import android.view.View
import com.eftimoff.bakingapp.R
import com.eftimoff.bakingapp.app.injection.AppComponent
import com.eftimoff.bakingapp.app.models.Recipe
import com.eftimoff.bakingapp.app.view.BaseFragment
import com.eftimoff.bakingapp.recipedetails.view.RecipeDetailsActivity
import com.eftimoff.bakingapp.recipelist.di.RecipeListModule
import com.eftimoff.bakingapp.recipelist.viewmodels.RecipeListViewModel
import com.eftimoff.bakingapp.recipelist.viewmodels.RecipeListViewModelFactory
import kotlinx.android.synthetic.main.fragment_recipe_list.*
import javax.inject.Inject


class RecipeListFragment : BaseFragment(), RecipeAdapter.Callback {

    @Inject
    lateinit var recipeListViewModelFactory: RecipeListViewModelFactory
    @Inject
    lateinit var recipeAdapter: RecipeAdapter
    @Inject
    lateinit var layoutManager: RecyclerView.LayoutManager

    lateinit var viewModel: RecipeListViewModel

    companion object {
        fun newInstance(): RecipeListFragment {
            return RecipeListFragment()
        }
    }

    override fun inject(appComponent: AppComponent) {
        appComponent.plus(RecipeListModule(this)).inject(this)
    }

    override fun getLayoutRes(): Int {
        return R.layout.fragment_recipe_list
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        recipeAdapter.setCallback(this)
        recipeList.adapter = recipeAdapter
        recipeList.layoutManager = layoutManager

        viewModel = ViewModelProviders.of(this, recipeListViewModelFactory).get(RecipeListViewModel::class.java)
        viewModel.getData().observe(this, Observer {
            if (it != null) {
                onRecipesSuccess(it)
            }
        })
    }

    fun onRecipesSuccess(recipes: List<Recipe>) {
        recipeAdapter.setRecipes(recipes)
    }

    override fun onRecipeClicked(recipe: Recipe) {
        RecipeDetailsActivity.start(context, recipe)
    }

}